pkg/runtime/generate.go

```diff
@@ -136,80 +159,124 @@ func (a *addonFile) genImports() []byte {
+			a.writeKubePutYAMLWithIndent(buf, o, 1)
+		}
+		if i != len(a.objects)-1 {
+			buf.WriteString("\n")
+		}
+		a.currentIndex++
+	}
+	return buf.Bytes()
+}
+
+func (a *addonFile) writeKubePutWithIndent(kubePut *bytes.Buffer, object k8sruntime.Object, indent int) []byte {
 	indent1 := bytes.Repeat([]byte(indentString), indent)
 	indent2 := bytes.Repeat([]byte(indentString), indent+1)
-	kubePut := bytes.NewBuffer([]byte{})
-	for i, object := range a.objects {
-		data := a.genDataWithIndent(reflect.ValueOf(object), indent+2)
+	data := a.genDataWithIndent(reflect.ValueOf(object), indent+2)
 
-		kubePut.Write(indent1)
-		kubePut.WriteString("kube.put(\n")
+	group := object.GetObjectKind().GroupVersionKind().Group
+	kind := object.GetObjectKind().GroupVersionKind().Kind
 
-		// name
-		kubePut.Write(indent2)
-		kubePut.WriteString("name=")
-		kubePut.WriteString(a.names[i])
-		kubePut.WriteString(",\n")
-
-		// api_group
-		apiGroup := object.GetObjectKind().GroupVersionKind().Group
-		if apiGroup != "" {
-			kubePut.Write(indent2)
-			kubePut.WriteString("api_group=\"" + apiGroup + "\",\n")
-		}
+	a.metaData[a.currentIndex].group = group
+	a.metaData[a.currentIndex].kind = kind
 
-		// data
-		kubePut.Write(indent2)
-		kubePut.WriteString("data=[")
-		kubePut.Write(data)
-		kubePut.WriteString("]\n")
+	kubePut.Write(indent1)
+	kubePut.WriteString("kube.put(\n")
 
-		kubePut.Write(indent1)
-		kubePut.WriteString(")\n")
+	// name
+	kubePut.Write(indent2)
+	kubePut.WriteString("name=\"" + a.metaData[a.currentIndex].name + "\",\n")
 
-		if i != len(a.objects)-1 {
-			kubePut.WriteString("\n")
-		}
+	// api_group
+	if group != "" {
+		kubePut.Write(indent2)
+		kubePut.WriteString("api_group=\"" + group + "\",\n")
 	}
+
+	// data
+	kubePut.Write(indent2)
+	kubePut.WriteString("data=[")
+	kubePut.Write(data)
+	kubePut.WriteString("]\n")
+
+	kubePut.Write(indent1)
+	kubePut.WriteString(")\n")
+
 	return kubePut.Bytes()
 }
 
+func (a *addonFile) writeKubePutYAMLWithIndent(kubePutYAML *bytes.Buffer, c unstructured.Unstructured, indent int) []byte {
+	indent1 := bytes.Repeat([]byte(indentString), indent)
+	indent2 := bytes.Repeat([]byte(indentString), indent+1)
+	name := c.GetName()
+	namespace := c.GetNamespace()
+	group := c.GroupVersionKind().Group
+	kind := c.GroupVersionKind().Kind
+
+	a.metaData[a.currentIndex] = metaData{
+		name:      name,
+		namespace: namespace,
+		group:     group,
+		kind:      kind,
+	}
+
+	kubePutYAML.Write(indent1)
+	kubePutYAML.WriteString("kube.put_yaml(\n")
+
+	// name
+	kubePutYAML.Write(indent2)
+	kubePutYAML.WriteString("name=\"" + name + "\",\n")
+
+	// namespace
+	if namespace != "" {
+		kubePutYAML.Write(indent2)
+		kubePutYAML.WriteString("namespace=\"" + namespace + "\",\n")
+	}
+
+	// data
+	kubePutYAML.Write(indent2)
+	kubePutYAML.WriteString(`data=["""`)
+	// error can be safely ignored here cause at this point we know it's a valid structure
+	j, _ := c.MarshalJSON()
+	j = bytes.TrimSpace(j)
+	kubePutYAML.Write(j)
```

Comment:
This is probably fine for now.

In the future, it would be super awesome if isopod could instead generate `struct(foo=bar, baz=qux).to_json()` for the `data` field of `kube.put_yaml()`. I mean, even for k8s.io objects, we could use this exact procedure to generate `kube.put_yaml()` statements that are technically correct translation, but not that different from vanilla yaml files. I think one of the true powers of Isopod is code reusability, where the users could take the generated  `struct(foo=bar, baz=qux).to_json()` and turn it into another function, for example, for Istio CRDs 
```python
def destination_rule(name, namespace, tls_mode="ISTIO_MUTUAL"):
    meta = metadata(name, namespace)
    dr = struct(
        apiVersion="networking.istio.io/v1alpha3",
        kind="DestinationRule",
        metadata=meta,
        spec=struct(
            host="{}.{}.svc.cluster.local".format(name, namespace),
            trafficPolicy=struct(
                tls=struct(
                    mode=tls_mode,
                ),
            ),
            subsets=[
                struct(
                    name="v1",
                    labels=struct(
                        version="v1",
                    ),
                ),
            ],
        ),
    )
    return dr.to_json()
```

Reply:
so should we have done all code generation as Starlark structs? would have saved us the imports and the types.

Reply:
For runtime objects that are not custom resources, we should still use `kube.put()` and generate code with the proto package imports, because typed k8s objects ensure their fields/members are named and typed (primitive types or another k8s type) correctly.

`kube.put_yaml()` was devised as a workaround specifically for custom resources, because k8s API does not support protos for custom resources yet. My previous comment was trying to make a point that 
```python
kube.put_yaml(
    name="foo",
    namespace="foo-ns",
    data=["""
apiVersion: example.com/v1
kind: ExampleResource
metadata:
  name: foo
  namespace: foo-ns
spec:
  # ...
"""],
)
```
is not a whole lot better than whatever templating scheme that the users used prior to Isopod. Using Starlark structs gets us closer to what using imported proto packages looks like.

Depends on your time estimate on this feature, I am okay with making a second PR later and prioritize your other sprint goals

Reply:
this is not in my sprint, just a hobby project 😊
Yeah if you're okay with it I think we should merge and we can add the Starlark rendering in the future. I created #51 for it.